sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2437,9 +2435,12 @@ class Dataset[T] private[sql](
    * @group typedrel
    * @since 1.6.0
    */
-  def coalesce(numPartitions: Int): Dataset[T] = withTypedPlan {
-    Repartition(numPartitions, shuffle = false, logicalPlan)
-  }
+  def coalesce(numPartitions: Int, partitionCoalescer: Option[PartitionCoalescer]): Dataset[T] =
+    withTypedPlan {
+      CoalesceLogical(numPartitions, partitionCoalescer, logicalPlan)
+    }
+
+  def coalesce(numPartitions: Int): Dataset[T] = coalesce(numPartitions, None)
```

Comment:
Please also add the function description, like what we did in the other functions in Dataset.scala?